### ISSUE TITLE
Keras `predict` with sigmoid output returns probabilities instead of labels

### ISSUE THREAD

**Issue by Kodiologist** [tensorflow/tensorflow#35876]
**System information**

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yee
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 19.10
- TensorFlow installed from (source or binary): PyPI wheel
- TensorFlow version: v2.1.0-rc2-17-ge5bf8de
- Python version: 3.7.5

**Describe the current behavior**

The `predict` method of a Keras model with a sigmoid activiation function for the output returns probabilities.

**Describe the expected behavior**

`predict` should return class indices or class labels, as in the case of softmax activation. The current behavior also seems to be redundant with the method `predict_proba`, which in turn doesn't seem to be documented, at least on https://www.tensorflow.org/api_docs/python/tf/keras/Model

#7287 is perhaps related.

**Code to reproduce the issue**

```python
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation

model = Sequential([
    Dense(10, activation = "relu"),
    Dense(1, activation = "sigmoid")])
model.compile(
    optimizer = "rmsprop",
    loss = "binary_crossentropy")
model.fit(
    [[1, 2], [1, 3], [1, 1], [2, 2], [2, 3]],
    [True, False, False, True, True])

print(model.predict([[1, 2], [1, 3], [1, 1]]))
```

**Comment by pandrey-fr** [comment]
There is no `predict_proba` method in the keras API, contrary to the scikit-learn one.

Thus, `predict` always returns the predicted probabilities, which you can easily transform into labels if you wish, either using `tf.argmax(prediction, axis=-1)` (for softmax activation) or, in your example case, `tf.greater(prediction, .5)` (provided you want to use a .5 threshold, which you could also adjust depending on your modelling context).

**Comment by Kodiologist** [comment]
> There is no `predict_proba` method in the keras API, contrary to the scikit-learn one.

`model.predict_proba` exists, has a docstring, seems to do the same thing as `predict` (at least for sigmoid activation), and doesn't seem to have a deprecation warning or anything.

> Thus, `predict` always returns the predicted probabilities

Not for softmax activation.

**Comment by pandrey-fr** [comment]
> `model.predict_proba` exists, has a docstring, seems to do the same thing as `predict` (at least for sigmoid activation), and doesn't seem to have a deprecation warning or anything.

Sorry, it appears that `Sequential` indeed provides with a `predict_proba` method; it is, however, not inherited from the base `Model` class, and thus not always available (hence my ignoring its existence - sorry). As `predict` comes from this base class, which does not have a `predict_proba`, it returns predicted probabilities (it it acted differently, it would be breaking the basic keras API).

Now, I do not understand why there is a `predict_proba` implemented as `Sequential` level... I will check the source code.

> Not for softmax activation.

I must insist that softmax does yield predicted probabilities.

```python
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation

model = Sequential([
    Dense(10, activation = "relu"),
    Dense(2, activation = "softmax")])
model.compile(
    optimizer = "rmsprop",
    loss = "sparse_categorical_crossentropy")

print(model.predict([[1, 2], [1, 3], [1, 1]]))
```

Output (in my case, which obviously will not yield the same values due to weights' randomness):
```
[[0.43271175 0.5672882 ]
 [0.3681818  0.63181823]
 [0.5206317  0.4793683 ]]
```

**Comment by pandrey-fr** [comment]
Okay, so [source code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/sequential.py) indicates that `Sequential` implements both the `predict_proba` method (that merely calls `predict` and raises an exception if there is any value outside of ]0, 1[) and a `predict_classes` one that basically does what I first indicated (using argmax or a .5 threshold) based on, once again, the outputs of `predict`.

**Comment by Kodiologist** [comment]
> I must insist that softmax does yield predicted probabilities.

I was getting all 1s when I called `model.predict` with softmax, but I had only 1 output node rather than 2 (`Dense(1, activation = "softmax")`), which in hindsight, you're probably not supposed to do, unlike for sigmoid activation. Perhaps there should be an error message for this. Thanks for showing me how to do it right.

**Comment by Kodiologist** [comment]
Perhaps `predict_proba` and `predict_classes` be added to `Model` or removed from `Sequential` for consistency.

**Comment by pandrey-fr** [comment]
> I was getting all 1s when I called model.predict with softmax, but I had only 1 output node

Well, since softmax balances out a vector of values so that they sum to 1, having a single node will make its output probability jump to 1 in any case, which is indeed an undesirable behavior but could understandably cause your confusion :)

> Perhaps `predict_proba` and `predict_classes` should be added to Model or removed from Sequential for consistency.

Let us wait for a tensorflower to pick up the issue and give his/her point of view on that question.
Imho, the issue with porting it everywhere is that the implemented methods are targeted at single output tensors, while `Model` allows for much more general behaviors. But having these methods in `Sequential` can indeed be confusing.

**Comment by oanush** [comment]
@Kodiologist ,
Any update on the issue ?Thanks!

**Comment by Kodiologist** [comment]
@oanush I don't understand. What sort of update are you hoping to receive from me?

**Comment by ymodak** [comment]
@Kodiologist Thanks for your issue. Perhaps  you may want to raise a feature request for adding `predict_proba` and `predict_classes` methods  for functional or subclassed models.
Closing this issue since `model.predict` method is behaving as intended. Thanks!

**Comment by Kodiologist** [comment]
@ymodak It's just as well that you didn't actually close the issue, since you might as well use this issue for the API and documentation complaints I mentioned earlier instead of opening a new issue for them.

**Comment by libphy** [comment]
Perhaps now the requested features are available: https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/wrappers/scikit_learn.py#L225

**Comment by ymodak** [comment]
Closing this issue since it's resolved. Thanks
